core/codegen/src/attribute/async_entry.rs

```diff
@@ -1,83 +1,134 @@
 }
 
-impl Kind {
-    // The name of the attribute, used for error messages
-    fn attr_name(&self) -> &'static str {
-        match self {
-            Kind::Main => "main",
-            Kind::Test => "async_test",
+struct Main;
+
+impl EntryAttr for Main {
+    const REQUIRES_ASYNC: bool = true;
+
+    fn function(f: &syn::ItemFn, block: &syn::Block) -> Result<TokenStream2> {
+        let (vis, mut sig) = (&f.vis, f.sig.clone());
+        if sig.ident != "main" {
+            return Err(Span::call_site()
+                .error("attribute can only be applied to `main` function")
+                .span_note(sig.span(), "this function must be `main`"));
         }
+
+        sig.asyncness = None;
+        Ok(quote_spanned!(block.span().into() => #vis #sig {
+            ::rocket::async_main(async move #block)
+        }))
     }
+}
 
-    // Attributes to decorate the generated function with
-    fn attrs(&self) -> Option<TokenStream2> {
-        match self {
-            Kind::Main => None,
-            Kind::Test => Some(quote!{ #[test] }),
-        }
+struct Test;
+
+impl EntryAttr for Test {
+    const REQUIRES_ASYNC: bool = true;
+
+    fn function(f: &syn::ItemFn, block: &syn::Block) -> Result<TokenStream2> {
+        let (vis, mut sig) = (&f.vis, f.sig.clone());
+        sig.asyncness = None;
+        Ok(quote_spanned!(block.span().into() => #[test] #vis #sig {
+            ::rocket::async_test(async move #block)
+        }))
     }
+}
+
+struct Launch;
 
-    // The path to the function to call
-    fn fn_path(&self) -> TokenStream2 {
-        match self {
-            Kind::Main => quote! { rocket :: async_main },
-            Kind::Test => quote! { rocket :: async_test },
+impl EntryAttr for Launch {
+    const REQUIRES_ASYNC: bool = false;
+
+    fn function(f: &syn::ItemFn, block: &syn::Block) -> Result<TokenStream2> {
+        if f.sig.ident == "main" {
+            return Err(Span::call_site()
+                .error("attribute cannot be applied to `main` function")
+                .note("this attribute generates a `main` function")
+                .span_note(f.sig.span(), "this function cannot be `main`"));
+        }
+
+        let ty = match &f.sig.output {
+            syn::ReturnType::Type(_, ty) => ty,
+            _ => return Err(Span::call_site()
+                .error("attribute can only be applied to functions that return a value")
+                .span_note(f.sig.span(), "this function must return a value"))
+        };
+
+        let (vis, mut sig) = (&f.vis, f.sig.clone());
+        sig.ident = syn::Ident::new("main", sig.ident.span());
+        sig.output = syn::ReturnType::Default;
+        sig.asyncness = None;
+
+        let rocket = quote_spanned!(ty.span().into() => {
+            let ___rocket: #ty = #block;
+            let ___rocket: ::rocket::Rocket = ___rocket;
+            ___rocket
+        });
+
+        if f.sig.asyncness.is_some() {
```

Comment:
The reason this is different is because we want compilation to fail if the user uses `await` in a non-`async` function, so the non-`async` version takes care not to wrap their code in `async`. This was necessary when we were converting the function instead of re-emitting, but I believe this may not longer be necessary since the error should be caught by Rust itself on the original function.

Reply:
Ah right, I missed that. I think the new version makes perfect sense, then.